### PR TITLE
Add Sphinx to autogenerate user-facing docs

### PR DESCRIPTION
**Pull request by sunank200**
Use [Sphinx](https://www.sphinx-doc.org/en/master/usage/quickstart.html) and [Sphinx auto-api](https://sphinx-autoapi.readthedocs.io/en/latest/) extension to autogenerate docs to achieve the following:

- Better discoverability of the available Async Operators, Sensors and Hooks
- Allows us to document the limitations of certain async operators
- Add better compatibility matrix of Airflow and astronomer-providers package.
- Better Document versioning if we host it to [ReadTheDocs](https://readthedocs.org/)

closes: #64

### CODE REVIEW

**Comment by kaxil** [comment]
Rebase on `main` for fixing the failing tests

File: docs/astronomer/amazon/aws/example_dags/example_s3/index.rst [hunk 1]
```diff
@@ -0,0 +1,50 @@
+:py:mod:`amazon.aws.example_dags.example_s3`
+============================================
+
```
Comment (kaxil) [inline docs/astronomer/amazon/aws/example_dags/example_s3/index.rst:3]: These files are auto-generated, they shouldn't be in git!

File: docs/conf.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+release = "1.0.0a1"
+
+
+# -- General configuration ---------------------------------------------------
+
+# Add any Sphinx extension module names here, as strings. They can be
+# extensions coming with Sphinx (named 'sphinx.ext.*') or your custom
+# ones.
+extensions = [
+    "autoapi.extension",
```
Comment (kaxil) [inline docs/conf.py:34]: https://readthedocs.org/projects/astronomer-providers/builds/16358761/ is failing.

You should add this dependency to `setup.cfg` and include it in a new extra: `doc` and install it for readthedocs.yml

File: docs/conf.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    "sphinx.ext.autodoc",
+    "sphinx.ext.viewcode",
+    "sphinx.ext.intersphinx",
+    "sphinx.ext.inheritance_diagram",
```
Comment (kaxil) [inline docs/conf.py:38]: Why are these required?

File: docs/astronomer/amazon/aws/example_dags/example_s3/index.rst (hunk 1 above)
Comment (sunank200) [reply docs/astronomer/amazon/aws/example_dags/example_s3/index.rst]: Removed this.

File: docs/conf.py (hunk 3 above)
Comment (sunank200) [reply docs/conf.py]: Removed it

**Review by kaxil — APPROVED** [review]

**Comment by kaxil** [comment]
I made some changes via the following commits @sunank200 

- https://github.com/astronomer/astronomer-providers/pull/114/commits/511e60ae754937f1d03fb17d19bdb5ae4d8d5361
- https://github.com/astronomer/astronomer-providers/pull/114/commits/8360dd13be154d5a1af6c8ad706d55e0aad6ca9d
- https://github.com/astronomer/astronomer-providers/pull/114/commits/c25232af95b200cc59c55966f112754dd14dbe81
